File: test/technology/sidb/simulation/analysis/test_energy_distribution.cpp
/*
 * Copyright (c) 2018 - 2023 Marcel Walter
 * Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

/**
 * @file
 * @brief Tests for `fiction/technology/sidb/simulation/analysis/energy_distribution.hpp`.
 * @author Jan Drewniok (Drewniok)
 * @author Willem Lambooy (wlambooy)
 * @author Marcel Walter (marcelwa)
 */

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>

#include <fiction/technology/sidb/charge_distribution.hpp>
#include <fiction/technology/sidb/lattice.hpp>
#include <fiction/technology/sidb/layout.hpp>
#include <fiction/technology/sidb/model/charge_state.hpp>
#include <fiction/technology/sidb/simulation/analysis/energy_distribution.hpp>
#include <fiction/technology/sidb/simulation/potential_landscape.hpp>
#include <fiction/technology/sidb/technology.hpp>

#include <vector>

using namespace fiction;
using namespace fiction::sidb;
using namespace fiction::sidb::model;
using namespace fiction::sidb::simulation;
using namespace fiction::sidb::simulation::analysis;

TEST_CASE("Test calculate_energy_distribution function", "[energy-distribution]")
{
    SECTION("one empty layout")
    {
        const layout                           lyt{};
        const potential_landscape              land{lyt};
        const std::vector<charge_distribution> all_cds{land.evaluate(charge_distribution{lyt})};

        auto result = calculate_energy_distribution(all_cds);
        CHECK(result.size() == 1);
        const auto state = result.get_nth_state(0);
        REQUIRE(state.has_value());
        if (state.has_value())
        {
            CHECK_THAT(state->electrostatic_potential_energy, Catch::Matchers::WithinAbs(0.0, 0.00001));
        }
    }

    SECTION("one layout with one SiDB placed")
    {
        layout lyt{};
        lyt.assign_cell_type({0, 0, 0}, sidb_technology::cell_type::NORMAL);

        const potential_landscape              land{lyt};
        const std::vector<charge_distribution> all_cds{land.evaluate(charge_distribution{lyt, charge_state::NEUTRAL})};

        auto result = calculate_energy_distribution(all_cds);
        CHECK(result.size() == 1);
        result.for_each([&](const auto& energy [[maybe_unused]], const auto& degeneracy) { CHECK(degeneracy == 1); });
    }

    SECTION("several layouts")
    {
        layout lyt{};
        lyt.assign_cell_type({10, 5, 0}, sidb_technology::cell_type::NORMAL);
        lyt.assign_cell_type({11, 5, 0}, sidb_technology::cell_type::NORMAL);
        lyt.assign_cell_type({12, 5, 0}, sidb_technology::cell_type::NORMAL);

        const potential_landscape land{lyt, simulation_parameters{}};

        const auto first = land.evaluate(charge_distribution{lyt, charge_state::NEUTRAL});

        charge_distribution second{lyt, charge_state::NEUTRAL};
        second.assign_charge_state({11, 5, 0}, charge_state::NEGATIVE);
        second = land.evaluate(second);

        const auto third = land.evaluate(charge_distribution{lyt, charge_state::NEGATIVE});

        const std::vector<charge_distribution> all_cds{first,  first,  first, second, second,
                                                       second, second, third, third,  third};

        const auto result = calculate_energy_distribution(all_cds);

        // three distinct distributions; the last two have an identical potential energy (degeneracy 2) which is
        // smaller than the one of the first
        const auto ground_state = result.get_nth_state(0);
        REQUIRE(ground_state.has_value());
        if (ground_state.has_value())
        {
            CHECK(ground_state->degeneracy == 2);
        }

        const auto excited_state = result.get_nth_state(1);
        REQUIRE(excited_state.has_value());
        if (excited_state.has_value())
        {
            CHECK(excited_state->degeneracy == 1);
        }
    }
}

File: bindings/mnt/pyfiction/test/technology/sidb/simulation/analysis/test_energy_distribution.py
# Copyright (c) 2018 - 2023 Marcel Walter
# Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Tests for ``calculate_energy_distribution``."""

from __future__ import annotations

from itertools import pairwise

import pytest

from mnt.pyfiction import (
    calculate_energy_distribution,
    charge_distribution,
    lattice,
    lattice_site,
    potential_landscape,
    sidb_charge_state,
    sidb_layout,
    sidb_simulation_parameters,
    sidb_technology,
)


@pytest.mark.parametrize(
    ("lat", "all_negative_energy"),
    [
        pytest.param(lattice.si_100_2x1(), 0.48066663155586997, id="100"),
        pytest.param(lattice.si_111_1x1(), 0.233980661373219, id="111"),
    ],
)
def test_three_sidbs(lat, all_negative_energy):
    layout = sidb_layout(lat)

    layout.assign_cell_type(lattice_site(0, 0, 1), sidb_technology.cell_type.NORMAL)
    layout.assign_cell_type(lattice_site(4, 0, 1), sidb_technology.cell_type.NORMAL)
    layout.assign_cell_type(lattice_site(6, 0, 1), sidb_technology.cell_type.NORMAL)

    landscape = potential_landscape(layout, sidb_simulation_parameters())

    cd1 = landscape.evaluate(charge_distribution(layout))  # all negative
    cd2 = landscape.evaluate(charge_distribution(layout, sidb_charge_state.NEUTRAL))  # all neutral
    cd3 = charge_distribution(layout)
    cd3.assign_charge_state(lattice_site(6, 0, 1), sidb_charge_state.NEUTRAL)  # only two SiDBs are negative
    cd3 = landscape.evaluate(cd3)
    cd4 = charge_distribution(layout)
    cd4.assign_charge_state(lattice_site(0, 0, 1), sidb_charge_state.NEUTRAL)
    cd4.assign_charge_state(lattice_site(4, 0, 1), sidb_charge_state.POSITIVE)
    cd4.assign_charge_state(lattice_site(6, 0, 1), sidb_charge_state.NEGATIVE)
    cd4 = landscape.evaluate(cd4)

    assert cd1.energy() == pytest.approx(all_negative_energy, abs=1e-7)
    assert cd2.energy() == pytest.approx(0.0, abs=1e-7)
    assert 0.0 < cd3.energy() < cd1.energy()
    assert cd4.energy() < 0.0  # the positive SiDB attracts its negative neighbor

    distribution = calculate_energy_distribution([cd1, cd2, cd3, cd4])

    assert distribution.size() == 4
    assert distribution.min_energy() == pytest.approx(cd4.energy(), abs=1e-7)
    assert distribution.max_energy() == pytest.approx(cd1.energy(), abs=1e-7)

    states = [distribution.get_nth_state(i) for i in range(distribution.size())]
    assert all(s.degeneracy == 1 for s in states)
    assert all(
        lower.electrostatic_potential_energy < higher.electrostatic_potential_energy
        for lower, higher in pairwise(states)
    )


def test_degenerate_states():
    layout = sidb_layout()
    layout.assign_cell_type(lattice_site(0, 0, 0), sidb_technology.cell_type.NORMAL)
    layout.assign_cell_type(lattice_site(10, 0, 0), sidb_technology.cell_type.NORMAL)

    left = charge_distribution(layout, sidb_charge_state.NEUTRAL)
    left.assign_charge_state(lattice_site(0, 0, 0), sidb_charge_state.NEGATIVE)
    left.assign_energy(0.5)
    right = charge_distribution(layout, sidb_charge_state.NEUTRAL)
    right.assign_charge_state(lattice_site(10, 0, 0), sidb_charge_state.NEGATIVE)
    right.assign_energy(0.5)
    excited = charge_distribution(layout)
    excited.assign_energy(1.0)

    distribution = calculate_energy_distribution([excited, left, right])

    assert distribution.size() == 2
    assert distribution.get_nth_state(0).electrostatic_potential_energy == pytest.approx(0.5)
    assert distribution.get_nth_state(0).degeneracy == 2
    assert distribution.get_nth_state(1).degeneracy == 1
